### ISSUE TITLE
[Windows] Link error with SWIG_PYTHON_INTERPRETER_NO_DEBUG

### ISSUE THREAD

**Issue by sgallou** [swig/swig#325]
Hi all,

I'm using Swig to generate wrappers for Python. To avoid rebuilding all Python, I use the binaries version, provided as Windows installer. This installer doesn't provide Pyhton27_d.lib, the debug version of Python Library.
So, to build the wrapper in debug configuration, I had to define SWIG_PYTHON_INTERPRETER_NO_DEBUG.
It builds nice, until I want to use std::vector. In this case, I got this link error :+1: 
error LNK2019: unresolved external symbol __imp___invalid_parameter_noinfo_noreturn referenced in function `public: double & __thiscall std::vector<double,class std::allocator<double> >::operator[](unsigned int)` (`??A?$vector@NV?$allocator@N@std@@@std@@QAEAANI@Z`)

I found a discussion, and a workaround, about this problem here : http://thread.gmane.org/gmane.comp.science.paraview.user/10956/focus=11019

I test it modifying the Swig generated file, and it works for me.
But it's a "not very nice" workaround, and temporary (each new Swig execution erase the modification).

Is there a way to solve this problem properly ?

Thanks,
Sébastien

**Comment by wsfulton** [comment]
What exactly did you do to work around it?

**Comment by sgallou** [comment]
I add includes to these system headers just before the #undef DEBUG : 
basetsd.h
assert.h
ctype.h
errno.h
io.h
math.h
sal.h
stdarg.h
stddef.h
stdio.h
stdlib.h
string.h
sys/stat.h
time.h
wchar.h

To workaround the erasing of these modifications when running SWIG, I forced these includes for the .cxx generated file.

Awful...

**Comment by wsfulton** [comment]
Add them to the top of your wrapper file by putting them inside a %begin block. A patch to the Python.html documentation with this hack and the reason for it would be appreciated.

**Comment by alexxlzhou** [comment]
I got a similar error with SWIG-3.0.5. sgallou solution (when put in %begin block) worked for me. Thanks! 
Hopefully there will be more elegant fix in the future.

**Comment by f1af** [comment]
%begin %{
	#undef  _DEBUG
	#define _ALLOW_ITERATOR_DEBUG_LEVEL_MISMATCH 
%}

**Comment by f1af** [comment]
%begin %{
	#define _CRT_SECURE_INVALID_PARAMETER(expr)
%}

**Comment by sgallou** [comment]
Hi f1af,

I tried your solution, but doesn't work for me. When compiling my solution
in debug (so try to link with Python release library), I got this error :
error LNK2038: mismatch detected for 'RuntimeLibrary': value
'MDd_DynamicDebug' doesn't match value 'MD_DynamicRelease' in...

sgallou



2015-11-17 19:48 GMT+01:00 f1af <notifications@github.com>:

**Comment by f1af** [comment]
Try with _DEBUG and _ALLOW_ITERATOR_DEBUG_LEVEL_MISMATCH for simple halt error message  about _ITERATOR_DEBUG_LEVEL (somsing like that _ITERATOR_DEBUG_LEVEL = 2 missmatch _ITERATOR_DEBUG_LEVEL = 0). Yore solution must compiling normaly, but u crash when try read any vector item.

**Comment by wsfulton** [comment]
Another solution mentioned to me is to just ```#include <crtdefs.h>``` before the ```#undef _DEBUG```. This is because crtdefs.h contains the code causing the missing symbol error:

```
#ifdef _DEBUG
 _CRTIMP void __cdecl _invalid_parameter(_In_opt_z_ const wchar_t *, _In_opt_z_ const wchar_t *, _In_opt_z_ const wchar_t *, unsigned int, uintptr_t);
#else
 _CRTIMP void __cdecl _invalid_parameter_noinfo(void);
 _CRTIMP __declspec(noreturn) void __cdecl _invalid_parameter_noinfo_noreturn(void);
#endif
```
Presumably this should be done before any other headers use crtdefs.h, so use:
```
%begin %{
#if defined(_DEBUG) && defined(SWIG_PYTHON_INTERPRETER_NO_DEBUG)
#include <crtdefs.h>
#endif
%}
```

**Comment by sgallou** [comment]
Thanks a lot William, it works for me, and this is a nicer workaround.

**Comment by ojwb** [comment]
Can this ticket be closed?  If not, what actually needs doing - e.g. should we be documenting one/some of these workarounds somewhere appropriate?

**Comment by sgallou** [comment]
Hi ojwb,

the previously mentioned solution is just a workaround, it would be nice if defining SWIG_PYTHON_INTERPRETER_NO_DEBUG do all the job.

**Comment by wsfulton** [comment]
I suggest we only add ```#include <crtdefs.h>``` if the test-suite passes all tests for debug builds. I'd like to see a patch that runs the python test-suite on Appveyor that has debug builds turned on - https://github.com/swig/swig/blob/master/appveyor.yml .

**Comment by blowekamp** [comment]
I am just dealing with this issue in my project. I would much prefer if this fix/improvement could be integrated into upstream SWIG, so it does not need to be duplicated in other SWIG/python projects.

I just followed what has been done in VTK/pyQt

SimpleITK/SimpleITK#172

**Comment by vadz** [comment]
FWIW I think this one should be simply closed because Python 2.7 is obsolete and Python 3 provides debug binaries too, so there is no problem when using it.

**Comment by ojwb** [comment]
If this is Python2 only I agree it makes sense to close as "wontfix" at this point.

If anyone affected really thinks SWIG including a workaround is still useful, please open a PR with the proposed change.  Note wsfulton's requirement for CI coverage above.
